### PR TITLE
Cost Voting Contract

### PR DESCRIPTION
**Pull request by reedrosenbluth**
Boot contract for voting on cost functions

Issue: https://github.com/blockstack/stacks-blockchain/issues/1894

### CODE REVIEW

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [inline src/chainstate/stacks/boot/cost-voting.clar:16]: Can you add another `define-map` for "confirmed" proposals and maps for tracking entries in it?

```scheme
(define-map confirmed-cost-functions
   ((cost-function-id uint))
   ((confirmed-function 
      (optional  
        {  function-contract: principal,
           function-name: (string-ascii 50),
           cost-function-contract: principal,
           cost-function-name: (string-ascii 50),
           confirmed-height: uint })))
(define-map functions-to-confirmed-ids
   ((function-contract principal) (function-name (string-ascii 50))
   ((cost-function-id uint)))
(define-data-var cost-function-max-id uint)
```

These data structures would allow us to iterate over the currently defined cost functions by reading from the map from `(0, cost-function-max-id)`. The reason the "optional" is required is to be able to delete a cost-function in O(1) by storing a `none`. The reason we need the `functions-to-confirmed-ids` map is that we'll want to check whether or not a definition already exists when confirming a proposal, and if so, overwrite rather than add a new entry.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [inline src/chainstate/stacks/boot/cost-voting.clar:99]: As currently written, these method stubs won't pass validation -- can you update the stubs to return like `(err 1)` or something along those lines? That way, we can work on integrating the rest of the cost updating code in the stacks-node with this API while the voting contract itself is being implemented.

Comment (reedrosenbluth) [reply src/chainstate/stacks/boot/cost-voting.clar:99]: yup!

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [inline src/chainstate/stacks/boot/cost-voting.clar:14]: Is `cost-function-max-id` used for anything? If not, can you remove the data-var definition?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [inline src/chainstate/stacks/boot/cost-voting.clar:155]: I think you need to add a data-map like `(define-data-map excercised-vetos ((proposol-id uint) (veto-height uint)) ((vetoed bool)))` which you can use to ensure that at any given stacks block height, at most one veto has been issued per proposal.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [inline src/chainstate/stacks/boot/cost-voting.clar:137]: Is `withdraw-after-votes` necessary? Could a voter just use `withdraw-votes` if we removed the `expiration-block-height` check?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (reedrosenbluth) [reply src/chainstate/stacks/boot/cost-voting.clar:14]: I don't think it's currently being used. I think I may have pasted it in from your branch. Is there a reason we might want to have a max-id?

Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:14]: I don't think it's necessary.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (reedrosenbluth) [reply src/chainstate/stacks/boot/cost-voting.clar:137]: I don't think it's necessary. Will make this change. I think we'll want `withdraw-votes` to behave slightly differently if the proposal is expired or not though. If the proposal is expired, we can keep the vote count the same (to preserve a record of the vote)

**Review by kantai — COMMENTED** [review]
This looks good to me -- I think that there are two changes that should be made before this PR can be approved:

1. Make it so that vetos can't be applied by a miner multiple times in the same block.
2. Get rid of `withdraw-after-votes` and just allow voters to use `withdraw-votes`

**Review by kantai — APPROVED** [review]
LGTM!

I also think that you should change the base of the PR to `next` -- this PR won't break any behavior in `next`, so I don't think the `next-costs` branch is really required anymore.

I think we may want to tweak the veto behavior (so that there's something like a "veto period" after the vote is confirmed, i.e., a proposal receives the required number of STX votes, gets "stacks holder confirmed", and then there's a period of ~100 blocks or so, when a veto could be applied by miners. But I think addressing that can be done in another PR -- I'll open an issue for this.

File: src/chainstate/stacks/boot/contract_tests.rs
Comment (jcnelson) [inline src/chainstate/stacks/boot/contract_tests.rs:1145]: Want to address this before merge?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:28]: Not a bug, but can you use consistent tuple notation?  I see you use `{ }` below.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:47]: Wait, aren't function names 128 bytes long in the proposal maps above?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:54]: Duplicate map definition.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:105]: Slight optimization: you can load `expiration-block-height` first, and error out with `ERR_NO_SUCH_PROPOSAL` without having to load `cur-votes` and `cur-principal-votes`.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:115]: The `err` case should be unreachable, because `amount > u0`.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:132]: Slight optimization -- you can remove `(sender tx-sender)` and just use `tx-sender`.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:28]: I think that until #2115 is merged, this syntax needs to be used in `define-map`

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:157]: @kantai I remember a conversation we had in the past about certain PoX-derived state, like the consensus hash, being unsafe to expose to the Clarity VM because it could cause the same block to have two different interpretations (https://github.com/blockstack/stacks-blockchain/issues/1805).  Does this concern still apply here, since a block's miner is derived from PoX state (i.e. which block wins sortition is determined in part from the PoX bitvector)?  I think my fix to #1805 removes this concern, but I just want to make sure.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:162]: Shouldn't this be `>=`?  You have `(asserts! (< burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))` earlier.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:167]: I'm a little confused why `veto-height` is the _Stacks_ block height, and not the burn block height (which seems to be what you're using to track when votes happen).  Can you briefly explain the rationale?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:184]: `block-height` is the Stacks height, not the Bitcoin block height.  Is that what you mean here?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:93]: I'm curious why `u2016` is a hard-coded value?  Do we want to allow proposals to have variable voting periods (IIRC Bitcoin has allowed voting periods for up to a whole year, for example)?  If not, then `u2016` should be a constant in the code, since you use it in multiple places.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:188]: You might want to make these constants, and document them.  Is this documentation correct?

* "A proposal must voted for by at least 20% of the total liquid STX"
* "A proposal must be vetoed by no more than 20% of the mined blocks since the proposal was created"

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:167]: This map is tracking whether or not this particular miner has exercised their veto in this block -- so the miner here is identified using the stacks block height. This map is to prevent the same miner from exercising a veto twice in the same block.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:157]: I don't think that concern applies to the winning block -- if a block is ever being evaluated, that means that block is the winner -- so data from that block is always safe to expose (e.g., who the miner is and how much they committed).

File: src/chainstate/stacks/boot/contract_tests.rs
Comment (reedrosenbluth) [reply src/chainstate/stacks/boot/contract_tests.rs:1145]: Yup working on this now!

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (reedrosenbluth) [reply src/chainstate/stacks/boot/cost-voting.clar:115]: Right, I wasn't quite sure how to handle this. Is the best practice to have an error const called `ERR_UNREACHABLE` and use that?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (lgalabru) [reply src/chainstate/stacks/boot/cost-voting.clar:28]: Done!

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [reply src/chainstate/stacks/boot/cost-voting.clar:130]: That's what we've been doing -- it indicates a pretty severe bug (and may even be considered grounds for triggering a node crash if it occurs in the boot code).

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:185]: Typo -- miners don't have hash power in PoX/PoB.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:223]: I'm a little confused about these two methods -- `confirm-votes` and `confirm-miners`.  It seems that somebody (can be anybody) must call both `confirm-votes` and, if it success, proceed to then call `confirm-miners`.  `confirm-votes` can be called as soon as there are enough tokens voted, but no later than 2016 blocks after the proposal was created, right?  Once the 2016 blocks have passed, miners then have a 1008-block window in which they can veto a proposal.  If they do not do so, then at any time after the 2016 + 1008 block window, someone (anyone) can call `confirm-miners` to activate the proposal.  Is my understanding here correct?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:130]: You can also use `unwrap-panic` which will cause the whole transaction to abort.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:223]: Yeah -- that's the idea. A proposal goes through two phases:

Phase 1: STX holder voting. If the proposal isn't confirmed before expiration block, then the proposal fails. Confirmation requires crossing a threshold of the liquid stx.
Phase 2: Miner vetoing. Once a proposal is stx-voter-confirmed, then miners are able to veto the proposal. Vetos must be applied before the veto period ends. Once the veto period ends, a proposal can be confirmed if fewer than a threshold of vetos were applied.

Splitting this into phases makes it clear to miners when a proposal has gained enough support that it needs to be considered for vetoing. It also limits the amount of time that a miner needs to be issuing veto votes.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [inline src/chainstate/stacks/boot/cost-voting.clar:227]: I think the expiration-block-height here should be `(+ VETO_LENGTH burn-block-height)`, such that the veto period begins immediately once the proposal is "vote-confirmed".

File: src/chainstate/stacks/boot/contract_tests.rs
Comment (jcnelson) [inline src/chainstate/stacks/boot/contract_tests.rs:1199]: Can you make this bound tighter?  Miners have exactly 1008 blocks to veto.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:111]: Just want to confirm -- if somehow the chain manages to confirm a cost function that doesn't exist or is malformed (e.g. it does I/O), the cost-assessment logic will fall back to the last valid cost-function for the given function here?

Also, what happens if `function-contract` and `function-name` don't correspond to a real function?  Will the cost function still apply to it even if the function is created later?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [inline src/chainstate/stacks/boot/cost-voting.clar:144]: Can you briefly explain why there's a fungible voting token?  I can understand the need to ensure that STX deposits and withdrawals are necessary, but why is a data map insufficient for that?

Also, `unwrap-panic` can be used in place of having a `(err ERR_UNREACHABLE)` if you want to have the whole transaction abort.  A good reason to use an `ERR_UNREACHABLE`-like error is if you want to trigger some kind of error (or panic) in the VM itself, whereby the VM intercepts the result and checks the error message.  Is that going to happen here?

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:111]: Those questions are answered in #2143 -- invalid proposals are simply never "applied" to the cost tracker. A warning is emitted, and then they aren't considered again.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (kantai) [reply src/chainstate/stacks/boot/cost-voting.clar:144]: The FT allows users to apply post-conditions to prevent inadvertent withdrawals.

Comment (jcnelson) [reply src/chainstate/stacks/boot/cost-voting.clar:144]: Also, having a dedicated voting token allows you to prevent surreptitious withdrawals by using the "deny" mode in post-conditions -- unless you intended the voting token to move, any attempt to withdraw your STX from the vote will cause the transaction to abort.

File: src/chainstate/stacks/boot/cost-voting.clar
Comment (jcnelson) [reply src/chainstate/stacks/boot/cost-voting.clar:111]: Thanks!

**Review by jcnelson — APPROVED** [review]
LGTM!  Just make sure the tests are all passing before merge.

**Review by kantai — APPROVED** [review]
Great work! This is good to merge once the tests pass!

**Pull request merged**
